File: HoTT-MdlViewer-JavaFX/src/main/java/gde/mdl/ui/TreeFileInfo.java
package gde.mdl.ui;

import java.util.List;

import gde.mdl.messages.Messages;
import gde.mdl.ui.background.LoadFromSdCardTask;
import gde.model.serial.FileInfo;
import gde.model.serial.FileType;
import javafx.beans.property.SimpleStringProperty;
import javafx.beans.property.StringProperty;
import javafx.scene.Node;
import javafx.scene.control.TreeItem;
import javafx.scene.image.Image;
import javafx.scene.image.ImageView;

public class TreeFileInfo extends TreeItem<String> {
    public static final Image FILE = LoadFile("/File.gif");
    public static final Image OPEN_FOLDER = LoadFile("/Folder_open.gif");
    public static final Image CLOSED_FOLDER = LoadFile("/Folder_closed.gif");
    public static final Image ROOT_FOLDER = LoadFile("/Root.gif");

    private static Image LoadFile(final String fileName) {
        return new Image(SelectFromSdCard.class.getResource(fileName).toString());
    }

    private final FileInfo fileInfo;
    private final StringProperty portNameProperty;
    private final Node view;

    public TreeFileInfo(final Node view, final FileInfo info) {
        // retrieve value form fileInfo
        this(view, info.getName(), info);
    }

    public TreeFileInfo(final Node view, final String value) {
        // root node without fileInfo
        this(view, value, null);

    }

    public TreeFileInfo(final Node view, final String value, final FileInfo fileInfo) {
        super(value);
        this.view = view;
        this.fileInfo = fileInfo;
        portNameProperty = new SimpleStringProperty(this, "portName");

        // update treeitem's children on expand
        expandedProperty().addListener(e -> update());

        // update treeitem's children when portName changes
        portNameProperty.addListener((p, o, n) -> update());

        setExpanded(false);

        if (fileInfo == null) {
            setGraphic(new ImageView(ROOT_FOLDER));
            getChildren().add(new TreeItem<>(Messages.getString("SelectFromTransmitter.loading")));
        } else if (fileInfo.getType() == FileType.Dir) {
            setGraphic(new ImageView(CLOSED_FOLDER));
            getChildren().add(new TreeItem<>(Messages.getString("SelectFromTransmitter.loading")));
        } else
            setGraphic(new ImageView(FILE));
    }

    public FileInfo getFileInfo() {
        return fileInfo;
    }

    /**
     * Get port name from portName property or parent tree item.
     *
     * @return
     */
    public String getPortName() {
        if (getParent() != null)
            return ((TreeFileInfo) getParent()).getPortName();
        else
            return portNameProperty.get();
    }

    public StringProperty portNameProperty() {
        return portNameProperty;
    }

    public void setPortName(final String portName) {
        portNameProperty.set(portName);
    }

    /**
     * Update children of this tree item in a background task.
     */
    @SuppressWarnings("unchecked")
    void update() {
        if (isExpanded()) {
            LoadFromSdCardTask task = null;

            if (fileInfo == null)
                // root node
                task = new LoadFromSdCardTask(view, getPortName(), "/");
            else if (fileInfo.getType() == FileType.Dir) // dir node
                task = new LoadFromSdCardTask(view, getPortName(), fileInfo.getPath());

            if (task != null) {
                task.setOnSucceeded(e -> {
                    if (fileInfo != null) setGraphic(new ImageView(OPEN_FOLDER));
                    getChildren().clear();
                    getChildren().addAll((List<TreeItem<String>>) e.getSource().getValue());
                });

                task.start();
            }
        } else {
            if (fileInfo != null) setGraphic(new ImageView(CLOSED_FOLDER));

            // add loading pseudo child
            getChildren().clear();
            getChildren().add(new TreeItem<>(Messages.getString("SelectFromTransmitter.loading")));
        }
    }
}

File: HoTT-MdlViewer-JavaFX/src/main/java/de/treichels/hott/mdlviewer/javafx/TreeFileInfo.kt
/**
 * HoTT Transmitter Config Copyright (C) 2013 Oliver Treichel
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software
 * Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
 * A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <http:></http:>//www.gnu.org/licenses/>.
 */
package de.treichels.hott.mdlviewer.javafx

import de.treichels.hott.decoder.HoTTSerialPort
import de.treichels.hott.model.serial.FileInfo
import de.treichels.hott.model.serial.FileType
import javafx.scene.control.TreeItem
import javafx.scene.control.TreeView
import javafx.scene.image.Image
import javafx.scene.image.ImageView
import tornadofx.*
import tornadofx.FX.Companion.messages


class TreeFileInfo(name: String, private val treeView: TreeView<String>, private val serialPort: HoTTSerialPort, internal val fileInfo: FileInfo? = null) : TreeItem<String>(name) {
    companion object {
        private fun image(fileName: String): Image = Image(TreeFileInfo::class.java.getResource(fileName).toString())

        private val fileImage = image("/File.gif")
        private val openFolderImage = image("/Folder_open.gif")
        private val closedFolderImage = image("/Folder_closed.gif")
        private val rootFolderImage = image("/Root.gif")
    }

    // helper
    private val isRoot = fileInfo == null
    private val isDir = fileInfo?.type == FileType.Dir
    private val isFile = fileInfo?.type == FileType.File

    internal val isReady
        get() = fileInfo != null && isFile && value.endsWith(".mdl") && fileInfo.size <= 0x3000 && fileInfo.size >= 0x2000

    init {
        isExpanded = false

        // update treeItem children on expand
        expandedProperty().addListener { _ -> update() }

        when {
            isRoot -> {
                graphic = ImageView(rootFolderImage)
            }

            isDir -> {
                graphic = ImageView(closedFolderImage)
            }

            isFile -> {
                graphic = ImageView(fileImage)
            }
        }
    }

    private fun loading() {
        // add loading pseudo child
        children.clear()
        children.add(TreeItem(messages["loading"]))
    }

    /**
     * Update children of this tree item in a background task.
     */
    private fun update() {
        when {
            isRoot -> {
                if (isExpanded) {
                    loadFromSdCard("/")
                } else {
                    loading()
                }
            }

            isDir -> {
                if (isExpanded) {
                    graphic = ImageView(openFolderImage)
                    loadFromSdCard(fileInfo!!.path)
                } else {
                    graphic = ImageView(closedFolderImage)
                    loading()
                }
            }
        }
    }

    private fun loadFromSdCard(path: String) {
        treeView.runAsyncWithOverlay {
            serialPort.use { p ->
                p.open()
                p.listDir(path).map { name ->
                    p.getFileInfo(name)
                }.map { info ->
                    TreeFileInfo(info.name, treeView, serialPort, info).apply {
                        if (isDir) loading()
                    }
                }
            }
        }.success { list ->
            children.clear()
            children.addAll(list)
        }
    }
}
